tensornetwork/block_sparse/utils.py

```diff
@@ -133,6 +136,216 @@ def get_dtype(itemsize):
+    if array.ndim > 1:
+      out[0] = out[0][:, 0:array.shape[1]]
+    if array.ndim == 1:
+      out[0] = np.squeeze(out[0])
+    if return_inverse and not return_index:
+      out[1] = out[1].astype(label_dtype)
+    elif return_inverse and return_index:
+      out[2] = out[2].astype(label_dtype)
+    out[0] = np.ascontiguousarray(out[0])
+
+  else:
+    if _return_index:
+      out = expand(res[0], array.dtype)
+    else:
+      out = expand(res, array.dtype)
+    if array.ndim > 1:
+      out = np.ascontiguousarray(out[:, 0:array.shape[1]])
+    if array.ndim == 1:
+      out = np.squeeze(out)
+
+  return out
+
+
+def intersect_new(A: np.ndarray,
+                  B: np.ndarray,
+                  axis=0,
+                  assume_unique=False,
+                  return_indices=False) -> Any:
+  """
+  Extends numpy's intersect1d to find the row or column-wise intersection of
+  two 2d arrays. Takes identical input to numpy intersect1d.
+  Args:
+    A, B (np.ndarray): arrays of matching widths and datatypes
+  Returns:
+    ndarray: sorted 1D array of common rows/cols between the input arrays
+    ndarray: the indices of the first occurrences of the common values in A.
+      Only provided if return_indices is True.
+    ndarray: the indices of the first occurrences of the common values in B.
+      Only provided if return_indices is True.
+  """
+  if A.dtype != B.dtype:
+    raise ValueError(f"array dtypes must macht to intersect, "
+                     f"found `A.dtype = {A.dtype}`, `B.dtype = {B.dtype}`")
+  if axis not in (0, 1):
+    raise NotImplementedError(
+        "intersection can only be performed on first or second axis")
+
+  if A.ndim != B.ndim:
+    raise ValueError("array ndims must match to intersect")
+
+  if axis == 1:
+    if A.shape[0] != B.shape[0]:
+      raise ValueError("array heights must match to intersect on second axis")
+
+    out = intersect_new(
+        A.T,
+        B.T,
+        axis=0,
+        assume_unique=assume_unique,
+        return_indices=return_indices)
+    if return_indices:
+      return np.ascontiguousarray(out[0].T), out[1], out[2]
+    return np.ascontiguousarray(out.T)
+
+  if A.ndim > 1 and A.shape[1] != B.shape[1]:
+    raise ValueError("array widths must match to intersect on first axis")
+  collapsed_A = collapse(A)
+  collapsed_B = collapse(B)
+
+  if collapsed_A.ndim > 1:
+    # arrays were not callapsable
+    return intersect(collapsed_A, collapsed_B, axis, assume_unique,
+                     return_indices)
+  if collapsed_A.dtype in (np.int8,
+                           np.int16) and collapsed_B.dtype in (np.int8,
+                                                               np.int16):
+    C, A_locs, B_locs = np.intersect1d(
+        collapsed_A,
+        collapsed_B,
+        assume_unique=assume_unique,
+        return_indices=True)
+    C = expand(C, A.dtype)
+    if A.ndim == 1:
+      C = np.squeeze(C)
+    if return_indices:
+      result = C, A_locs, B_locs
```

Comment:
There seems to be a lot of these `if return_indicies` in this function, and it affects the number of arguments returned. This seems like a recipe for issues in the future...

Reply:
True! The reason is to work around a `numpy` issue with causes `np.unique` to be about 5x slower for `np.int16` and `np.int8` dtypes when `return_indices=False` as opposed to `True`. `np.intersect1d has a similar issue, see https://github.com/numpy/numpy/issues/16923. I'll see if I can make the code clearer and more robust

Reply:
I tried to improve the code.  The issue is with`np.intersect1d`, which is called at the base of this function, returns different number of arguments based on the values of `return_indices`. Since the `intersect_new` wrapper adds an additional operation on the first returned value of this function, I need to check if `return_indices` has or has not been set (in addition to the dtype issue mentioned above).